include/revng/Support/ArtifactLocation.h

```diff
@@ -0,0 +1,162 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ADT/KeyedObjectTraits.h"
+#include "revng/ADT/StringTemplateParameter.h"
+#include "revng/Support/YAMLTraits.h"
+
+template<StringTemplateParameter... Names>
+struct ArtifactLocationNameHelper {
+  using ContainerType = std::array<std::string_view, sizeof...(Names)>;
+  static constexpr ContainerType Value{ Names... };
+  static constexpr size_t Count = sizeof...(Names);
+
+public:
+  template<StringTemplateParameter Name>
+  constexpr static std::optional<size_t> find() {
+    return findImpl<Name>();
+  }
+  template<StringTemplateParameter Name>
+  constexpr static bool contains() {
+    return find<Name>() != std::nullopt;
+  }
+
+private:
+  template<StringTemplateParameter Name, size_t CurrentIndex = 0>
+  constexpr static std::optional<size_t> findImpl() {
+    if constexpr (CurrentIndex >= Count)
+      return std::nullopt;
+    else if constexpr (Name == Value[CurrentIndex])
+      return CurrentIndex;
+    else
+      return findImpl<Name, CurrentIndex + 1>();
+  }
+};
+
+namespace pipeline {
+
+template<typename Names = ArtifactLocationNameHelper<>, typename... Types>
+class ArtifactLocationKind;
+
+} // namespace pipeline
+
+template<typename Names, typename... Types>
+struct ArtifactLocation : public std::tuple<Types...> {
+  using std::tuple<Types...>::tuple;
+  using NameHelper = Names;
+  friend pipeline::ArtifactLocationKind<Names, Types...>;
+
+  template<size_t Index>
+  using ElementType = std::tuple_element_t<Index, std::tuple<Types...>>;
+
+private:
+  std::string_view Kind;
```

Comment:
* `Kind` is quite an overloaded term in `revng`. What if we say that a location has a *name* and a set of *steps* or *arguments*?
* `std::string_view` is non-owning. Shall we go for `std::string` or `SmallString`?

Reply:
2. The non-owning nature is on-purpose. This value is only expected to ever be instantiated from a compile time literal - hence it points into a global lifetime literal. They are never changed or combiner or anything.

Moreover, the only reason this is not `const` is that it would break copy and move assignment.

Reply:
This is gone: this value is not `constexpr` and is called `RankName`.